Modules/posixmodule.c

```diff
@@ -1695,97 +1696,83 @@ static int
            get a handle to it. If the former, we need to return an error.
            If the latter, we can use attributes_from_dir. */
         DWORD lastError = GetLastError();
-        if (lastError != ERROR_ACCESS_DENIED &&
-            lastError != ERROR_SHARING_VIOLATION)
-            return -1;
-        /* Could not get attributes on open file. Fall back to
-           reading the directory. */
-        if (!attributes_from_dir(path, &info, &reparse_tag))
-            /* Very strange. This should not fail now */
-            return -1;
-        if (info.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
-            if (traverse) {
-                /* Should traverse, but could not open reparse point handle */
-                SetLastError(lastError);
+        if (lastError == ERROR_ACCESS_DENIED ||
+            lastError == ERROR_SHARING_VIOLATION) {
+            /* Could not get attributes on open file. Try reading
+               the directory. */
+            if (!attributes_from_dir(path, &info, &tagInfo.ReparseTag)) {
+                /* Very strange. This should not fail now */
                 return -1;
             }
+
+            /* Now info is initialized, fall through */
+        } else if (traverse && lastError == ERROR_CANT_ACCESS_FILE) {
+            /* Try again without traversal */
+            return win32_xstat_impl(path, result, FALSE);
+        } else {
+            return -1;
         }
     } else {
         if (!GetFileInformationByHandle(hFile, &info)) {
             CloseHandle(hFile);
             return -1;
         }
-        if (info.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
-            if (!win32_get_reparse_tag(hFile, &reparse_tag)) {
-                CloseHandle(hFile);
-                return -1;
-            }
-            /* Close the outer open file handle now that we're about to
-               reopen it with different flags. */
-            if (!CloseHandle(hFile))
-                return -1;
 
-            if (traverse) {
-                /* In order to call GetFinalPathNameByHandle we need to open
-                   the file without the reparse handling flag set. */
-                hFile2 = CreateFileW(
-                           path, FILE_READ_ATTRIBUTES, FILE_SHARE_READ,
-                           NULL, OPEN_EXISTING,
-                           FILE_ATTRIBUTE_NORMAL|FILE_FLAG_BACKUP_SEMANTICS,
-                           NULL);
-                if (hFile2 == INVALID_HANDLE_VALUE)
-                    return -1;
-
-                if (!get_target_path(hFile2, &target_path)) {
-                    CloseHandle(hFile2);
-                    return -1;
-                }
+        /* Get the reparse tag if we will need it */
+        if (info.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+            !GetFileInformationByHandleEx(hFile, FileAttributeTagInfo,
+                                          &tagInfo, sizeof(tagInfo))) {
+            tagInfo.ReparseTag = 0;
+        }
```

Comment:
`follow_symlinks` should only apply to actual symlinks. If `traverse` is false, then `GetFileInformationByHandleEx` : `FileAttributeTagInfo` should be checked before the (relatively expensive) `GetFileInformationByHandle` call, and only there because there's no point in checking it if `traverse` is true.  If it succeeds and `tagInfo.FileAttributes` contains `FILE_ATTRIBUTE_REPARSE_POINT` and the tag is not `IO_REPARSE_TAG_SYMLINK`, we have to close the handle and retstart with `win32_xstat_impl(path, result, TRUE)`.

Reply:
Well, not quite. Done like this it would also have to check here for `IO_REPARSE_TAG_APPEXECLINK`. It would be better to avoid hard coding that.  We need a value for `traverse` that indicates the call is handling `ERROR_CANT_ACCESS_FILE` and so shouldn't try again regardless of the tag. Either that or add another parameter for that case.

Reply:
We never retry because of the tag here - it is only used to set ST_IFLNK (which is then used to raise an error if we were supposed to follow links, which is the same as the previous behaviour). In `fileutils.c` I'm using the `IsReparseTagNameSurrogate(reparse_tag)` macro to set this flag now, so anything properly registered (which does not include `APPEXECLINK`) will appear as a link.

The `flags |= FILE_FLAG_OPEN_REPARSE_POINT` line higher up is what determines whether we are traversing links or not, and then we'll follow anything that the OS knows how to follow. If we're not traversing, the result is either success or an error - never a retry - so the value for `traverse` that indicates a retry is `FALSE`.

Reply:
`traverse` is from `follow_symlinks` and only applies to symlinks. It does not apply to other types of reparse points. If we've opened the reparse point to test for a symlink, we *must* reopen for all other types.